README.md
# mini-image-loader ![npm](https://img.shields.io/npm/v/mini-image-loader.svg?style=flat-square) ![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg?longCache=true&style=flat-square)

Small image processing loader for Webpack.

## Usage

```javascript
const imagePath = require("./duck.png?resize=900x&format=jpeg&quality=80");
```

Webpack config:

```javascript
{
  module: {
    rules: [{
      test: /\.(png|jpe?g|tiff|webp)$/,
      use: [
        {
          loader: "file-loader",
          options: {
            name: "[name].[hash:8].[ext]",
            outputPath: "img/"
          }
        },
        { loader: "mini-image-loader" },
      ]
    }]
  }
}
```

## Debugging

`DEBUG=mini-image-loader` enables verbose mode.

## API

### format

Force output to a given format.

Adjust the lossless compression quality with the `quality` property.

**Param**: `enum`, one of `png`, `jpeg`, `tiff`, `webp`.

```javascript
require("./duck.jpg?format=png");
require("./frog.png?format=webp");
```

### quality

Set quality for lossless compression (`jpeg` and `webp`).

**Param**: `integer`, between `1` and `100`.

```javascript
require("./frog.png?format=jpeg&quality=90");
```

### lossless

Set lossless compression for `webp`.

```javascript
require("./frog.png?format=webp&lossless");
```

### nearLossless

Set near lossless compression for `webp`.

```javascript
require("./frog.png?format=webp&nearLossless");
```

### resize

Resize image to `width` x `height`.

**Param**: `string`, following the `WIDTHxHEIGHT` pattern.

Examples:

```javascript
require("./duck.png?resize=120x240");
require("./cats.jpg?resize=900x");
require("./frog.png?resize=x300");
```

### crop

Crop the resized image to the exact size specified by `resize`.

The experimental strategy-based approach resizes so one dimension is at its target length then repeatedly ranks edge regions, discarding the edge with the lowest score based on the selected strategy.

- **entropy**: focus on the region with the highest Shannon entropy.
- **attention**: focus on the region with the highest luminance frequency, colour saturation and presence of skin tones.

**Param**: `string`, `north`, `northeast`, `east`, `southeast`, `south`, `southwest`, `west`, `northwest`, `center`, `centre`, `entropy` or `attention`.

Examples:

```javascript
require("./duck.png?resize=120x240&crop");
require("./cats.jpg?resize=900x100&crop=entropy");
require("./frog.png?resize=300x300&crop=south");
```

### gravity

Possible attributes of the optional sharp.gravity are north, northeast, east, southeast, south, southwest, west, northwest, center and centre.

### tint

Tint the image using the provided chroma while preserving the image luminance. An alpha channel may be present and will be unchanged by the operation.

**Param**: `string`, following the `rgb hex` format `#RRGGBB`.

Examples:

```javascript
require("./duck.png?tint=#BABACA");
```

### greyscale

Convert to 8-bit greyscale; 256 shades of grey. This is a linear operation. If the input image is in a non-linear colour space such as sRGB, use gamma with greyscale for the best results. By default the output image will be web-friendly sRGB and contain three (identical) color channels.

Examples:

```javascript
require("./duck.png?greyscale");
```

### grayscale

Alternative spelling of greyscale.

### blur

Blur the image.
When used without parameters, performs a fast, mild blur of the output image.
When a sigma is provided, performs a slower, more accurate Gaussian blur.

**Param**: `number?` a value between 0.3 and 1000 representing the sigma of the Gaussian mask, where sigma = 1 + radius / 2.

Examples:

```javascript
require("./duck.png?blur");
require("./cats.jpg?blur=1.2");
require("./frog.png?blur=200");
```

### gamma

Apply a gamma correction by reducing the encoding (darken) pre-resize at a factor of 1/gamma then increasing the encoding (brighten) post-resize at a factor of gamma. This can improve the perceived brightness of a resized image in non-linear colour spaces. JPEG and WebP input images will not take advantage of the shrink-on-load performance optimisation when applying a gamma correction.

**Param**: `number?`, between `1.0` and `3.0` (optional, default 2.2)

Examples:

```javascript
require("./duck.png?gamma");
require("./cats.jpg?gamma=1.2");
require("./frog.png?gamma=2.4");
```

### flip

Flip the image about the vertical Y axis.
The use of flip implies the removal of the EXIF Orientation tag, if any.

Examples:

```javascript
require("./duck.png?flip");
```

### flop

Flop the image about the horizontal X axis.
The use of flop implies the removal of the EXIF Orientation tag, if any.

Examples:

```javascript
require("./duck.png?flop");
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [ZeroVer](https://0ver.org/).

## Unreleased

## 0.1.0 - 2018-09-19

### Added

- `lossless`
- `nearLossless`

## 0.0.7 - 2018-09-19

### Added

- `crop`

## 0.0.6 - 2018-09-01

### Added

- `greyscale`
- `grayscale`

## 0.0.5 - 2018-09-01

### Added

- `tint`

## 0.0.4 - 2018-09-01

### Added

- `blur`

## 0.0.3 - 2018-09-01

### Added

- `flip` Flip the image about the vertical Y axis.
- `flop` Flip the image about the vertical X axis.

## 0.0.2 - 2018-09-01

### Added

- `gamma` corretion.

## 0.0.1 - 2018-08-31

### Added

- `format` change image format.
- `quality` set quality for lossless formats.
- `resize` change image width and height.


src/query-string.ts
import Ajv from "ajv";
import { parseQuery as luParseQuery } from "loader-utils";

import debug from "./debug";

export interface IQueryString {
  format?: string;
  quality?: number;

  resize?: string;

  tint?: string;
  greyscale?: boolean;
  grayscale?: boolean;

  blur?: number | boolean;
  gamma?: number | boolean;
  flip?: boolean;
  flop?: boolean;
}

const schema = {
  type: "object",
  additionalProperties: false,

  properties: {
    format: { type: "string", enum: ["png", "jpeg", "tiff", "webp"] },
    quality: { type: "number", minimum: 1, maximum: 100 },

    resize: { type: "string", pattern: "^\\d*x\\d*$" },

    tint: { type: "string", pattern: "^#[a-fA-F0-9]{6}$" },
    greyscale: { const: true },
    grayscale: { const: true },

    blur: {
      anyOf: [
        { const: true },
        { type: "number", minimum: 0.3, maximum: 1000 },
      ],
    },
    gamma: {
      anyOf: [
        { const: true },
        { type: "number", minimum: 1.0, maximum: 3.0 },
      ],
    },
    flip: { const: true },
    flop: { const: true },
  },

  dependencies: {
    quality: {
      properties: {
        format: {
          oneOf: [
            { const: "jpeg" },
            { const: "webp" },
          ],
        },
      },
    },
  },
};

const ajv = new Ajv({ allErrors: true, coerceTypes: true });
const validate = ajv.compile(schema);

export function parseQueryString(resourceQuery: string): IQueryString | Error {

  if (resourceQuery === "") {
    resourceQuery = "?";
  }

  const parsedQuery = luParseQuery(resourceQuery);
  debug("Parsed query string. %o", parsedQuery);

  const valid = validate(parsedQuery);

  if (!valid) {
    debug("Invalid query string %O", { parsedQuery, errors: validate.errors });
    return new Error(ajv.errorsText(validate.errors!, { dataVar: "QueryString" }));
  }

  debug("Valid query string %O", { parsedQuery });
  return parsedQuery as IQueryString;

}
